keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/plugins/ArtifactPublisher.kt

```diff
@@ -0,0 +1,53 @@
+package com.netflix.spinnaker.keel.api.plugins
+
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactStatus
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactType
+import com.netflix.spinnaker.keel.api.artifacts.BuildMetadata
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.artifacts.GitMetadata
+import com.netflix.spinnaker.keel.api.artifacts.SpinnakerArtifact
+import com.netflix.spinnaker.keel.api.artifacts.VersioningStrategy
+import com.netflix.spinnaker.keel.api.events.ArtifactEvent
+import com.netflix.spinnaker.keel.api.support.EventPublisher
+import com.netflix.spinnaker.kork.plugins.api.internal.SpinnakerExtensionPoint
+
+/**
+ * Keel plugin interface to be implemented by publishers of artifact information.
+ *
+ * The main job of an [ArtifactPublisher] is to detect new versions of artifacts, using
+ * whatever mechanism they choose (e.g. they could receive events from another system,
+ * or poll an artifact repository for artifact versions), and notify keel via the [publishArtifact]
+ * method, so that the artifact versions can be persisted and evaluated for promotion.
+ *
+ * The default implementation of [publishArtifact] simply publishes the event via the [EventPublisher],
+ * and should *not* be overridden by implementors.
+ */
+interface ArtifactPublisher<T : DeliveryArtifact> : SpinnakerExtensionPoint {
+  val eventPublisher: EventPublisher
+  val supportedArtifact: SupportedArtifact<T>
```

Comment:
I changed my mind on this and each `ArtifactPublisher` now supports a single type of artifact.